pynumaflow/mapper/_servicer/_async_servicer.py

```diff
@@ -0,0 +1,129 @@
+import asyncio
+from collections.abc import AsyncIterable
+
+from google.protobuf import empty_pb2 as _empty_pb2
+from pynumaflow.shared.asynciter import NonBlockingIterator
+
+from pynumaflow._constants import _LOGGER, STREAM_EOF
+from pynumaflow.mapper._dtypes import MapAsyncCallable, Datum
+from pynumaflow.proto.mapper import map_pb2, map_pb2_grpc
+from pynumaflow.shared.server import exit_on_error, handle_async_error
+from pynumaflow.types import NumaflowServicerContext
+
+
+class AsyncMapServicer(map_pb2_grpc.MapServicer):
+    """
+    This class is used to create a new grpc Async Map Servicer instance.
+    It implements the SyncMapServicer interface from the proto map.proto file.
+    Provides the functionality for the required rpc methods.
+    """
+
+    def __init__(
+        self,
+        handler: MapAsyncCallable,
+    ):
+        self.background_tasks = set()
+        self.__map_handler: MapAsyncCallable = handler
+
+    async def MapFn(
+        self,
+        request_iterator: AsyncIterable[map_pb2.MapRequest],
+        context: NumaflowServicerContext,
+    ) -> AsyncIterable[map_pb2.MapResponse]:
+        """
+        Applies a function to each datum element.
+        The pascal case function name comes from the proto map_pb2_grpc.py file.
+        """
+        # proto repeated field(keys) is of type google._upb._message.RepeatedScalarContainer
+        # we need to explicitly convert it to list
+        try:
+            # The first message to be received should be a valid handshake
+            req = await request_iterator.__anext__()
+            # check if it is a valid handshake req
+            if not (req.handshake and req.handshake.sot):
+                raise Exception("MapFn: expected handshake message")
```

Comment:
can we raise a more specific exception here?

Reply:
Sure! Can use a custom error type here